### PR TITLE
Convert to ES6 Class

### PR DESCRIPTION
**Pull request by sangm**
As someone who has never worked with `broccoli-persistent-filter` /
`broccoli-filter`, it was initially difficult to understand exactly what was going on.

IMO, if this was implemented as an ES6 class, it would be easier to understand.

Implemented a proxy to maintain backwards compatibility

### CODE REVIEW

**Comment by sangm** [comment]
Hm.. Didn't know we had to support node 4 :/

**Comment by Turbo87** [comment]
The correct way to instantiate this class is through the static `create()` method, using the constructor directly or even without using `new` is deprecated. Once we remove the deprecated code it should be easy to convert this to an ES6 class.

File: lib/index.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ const testGeneratorNames = Object.keys(testGenerators);
 
 const FACTORY_METHOD_USED = Symbol('create() factory method was used');
 
-function resolveInputDirectory(inputNode) {
```
Comment (rwjblue) [inline lib/index.js]: Let’s revert this change. Having a module scope function is pretty idiomatic IMHO.

File: lib/index.js [hunk 2]
```diff
@@ -29,214 +29,225 @@ function resolveInputDirectory(inputNode) {
...
- * @returns {EslintValidationFilter} Filter obconfig @constructor
- */
-function EslintValidationFilter(inputNode, options) {
-  if (!(this instanceof EslintValidationFilter)) {
-    return new EslintValidationFilter(inputNode, options);
-  }
+};
 
-  let _console = options.console || console;
+class _EslintValidationFilter extends Filter {
```
Comment (rwjblue) [inline lib/index.js]: Why underscore it?

File: lib/index.js [hunk 3]
```diff
@@ -29,214 +29,225 @@ function resolveInputDirectory(inputNode) {
...
+  }
+
+  static get testGenerators() {
     return testGeneratorNames.slice(0);
   }
+}
+
+// This is so that we remain backwards compatibile. Some people may invoke
+// without the `new` keyword (which is required for ES6 classes).
+const EslintValidationFilter = new Proxy(_EslintValidationFilter, {
```
Comment (rwjblue) [inline lib/index.js]: I don’t think we can rely on Proxy being available (would have to double check what Node 4 has to confirm)...

File: lib/index.js (hunk 2 above)
Comment (rwjblue) [reply lib/index.js]: Nevermind, I see

File: lib/index.js (hunk 3 above)
Comment (rwjblue) [reply lib/index.js]: I think you can get the same general result with a “normal” function here:

```
function EslintValidationFilter(inputNode, options) {
  if (this !== EslintValidationFilter) {
    console.warn(‘invoking broccoli-lint-eslint without new will be removed in a future version’);
  }

  return new _ESlintValidationFilter(inputNode, options);
}
```

Comment (rwjblue) [reply lib/index.js]: I believe this snippet (or a variant) will allow us to migrate to classes internally and keep newless Support...
